testindex/DependencyValidation.py

```diff
@@ -0,0 +1,126 @@
+        node = None
+        # Get the nodes from the networkx graph to iterate through
+        nodes = self._dependency_graph.nodes(data=True)
+        # Check if we need to get node by id or by name
+        # If neither is provided, raise exception
+        if not container_id and not container_name:
+            raise Exception("Either the id or the name of the container must be specified")
+        elif not container_id and container_name:
+            for item in nodes:
+                # Check the name label associated with the node to check if its the container_name we are looking for
+                if "name" in item[1] and container_name in item[1]["name"]:
+                    node = item
+                    break
+        # else if container id is provided, we just ignore container name
+        else:
+            for item in nodes:
+                # Check for matching id
+                if container_id == item[0]:
+                    node = item
+                    break
+        return node
+
+    def add_container(self, container_name, from_index=False, from_targetfile_baseimage=False):
+        """Checks if a node exists for container_name, and adds it if it does not exist"""
+        node = self.get_container_node(container_name=container_name)
+        if not node:
+            self._dependency_graph.add_node(self._node_number, name=container_name, from_index=from_index,
+                                            from_targetfile_baseimage=from_targetfile_baseimage)
+            self._node_number += 1
+        else:
+            fi = node[1]["from_index"]
+            ftb = node[1]["from_targetfile_baseimage"]
+            if not fi and from_index:
+                node[1]["from_index"] = True
+
+            if not ftb and from_targetfile_baseimage:
+                node[1]["from_targetfile_baseimage"] = True
+
+    def add_dependency(self, from_container, to_container):
+        """Add a dependency between the containers as a directed edge from one container to the other."""
+        # Get the nodes for the containers specified
+        from_node = self.get_container_node(container_name=from_container)
+        to_node = self.get_container_node(container_name=to_container)
+        # Check if nodes are available for both the containers, if not then, they should have been added already
+        if from_node and to_node:
+            # Extract he numbers which uniquely identify the node, so that an edge can be formed between them,
+            # if not already present
+            f = from_node[0]
+            t = to_node[0]
+            # Get the edges from the graph and check is an edge does not already exist for the pair, if not add it now.
+            edges = self._dependency_graph.edges()
+            if (f, t) not in edges:
+                self._dependency_graph.add_edge(f, t)
+
+        else:
+            raise Exception("Please add a both nodes before attempting to establish dependency")
```

Comment:
Add a check for finding the exact node which is not added to report better, also should we really need to `raise` here? see if returning other values can help.